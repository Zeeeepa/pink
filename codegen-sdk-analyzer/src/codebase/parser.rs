use codegen_sdk_ast::{Definitions, References, input::File};
#[cfg(feature = "serialization")]
use codegen_sdk_common::serialize::Cache;
use indicatif::{ProgressBar, ProgressStyle};

use super::discovery::{FilesToParse, log_languages};
use crate::{
    ParsedFile,
    database::{CodegenDatabase, Db},
    parser::parse_file,
};
fn execute_op_with_progress<Database: Db + ?Sized + 'static, T: Send + Sync>(
    db: &Database,
    files: FilesToParse,
    name: &str,
    op: fn(&Database, File) -> T,
) -> Vec<T> {
    let multi = db.multi_progress();
    let style = ProgressStyle::with_template(
        "[{elapsed_precise}] {wide_bar} {msg} [{per_sec}] [estimated time remaining: {eta}]",
    )
    .unwrap();
    let pg = multi.add(
        ProgressBar::new(files.files(db).len() as u64)
            .with_style(style)
            .with_message(name.to_string()),
    );
    let inputs = files
        .files(db)
        .into_iter()
        .map(|file| (&pg, file, op))
        .collect::<Vec<_>>();
    let results: Vec<T> = salsa::par_map(db, inputs, move |db, input| {
        let (pg, file, op) = input;
        let res = op(
            db,
            #[cfg(feature = "serialization")]
            &cache,
            file,
        );
        pg.inc(1);
        res
    });
    pg.finish();
    multi.remove(&pg);
    results
}
#[salsa::tracked]
fn parse_files_par(db: &dyn Db, files: FilesToParse) {
    let _: Vec<_> = execute_op_with_progress(db, files, "Parsing Files", |db, file| {
        parse_file(db, file);
    });
}
#[salsa::tracked]
fn parse_files_definitions_par(db: &dyn Db, files: FilesToParse) {
    let _: Vec<_> = execute_op_with_progress(db, files, "Parsing Definitions", |db, file| {
        let file = parse_file(db, file);
        if let Some(parsed) = file.file(db) {
            #[cfg(feature = "typescript")]
            if let ParsedFile::Typescript(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
            }
            #[cfg(feature = "python")]
            if let ParsedFile::Python(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
            }
        }
        ()
    });
}
pub fn parse_files<'db>(
    db: &'db CodegenDatabase,
    #[cfg(feature = "serialization")] cache: &'db Cache,
    files_to_parse: FilesToParse,
) -> () {
    rayon::ThreadPoolBuilder::new()
        .stack_size(1024 * 1024 * 1024 * 10)
        .build_global()
        .unwrap();
    log_languages();
    #[cfg(feature = "serialization")]
    let cache = Cache::new().unwrap();
    #[cfg(feature = "serialization")]
    let cached = get_cached_count(&cache, &files_to_parse);
    log::info!("Parsing {} files", files_to_parse.files(db).len());
    parse_files_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    log::info!("Parsing definitions");
    parse_files_definitions_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    #[cfg(feature = "serialization")]
    report_cached_count(cached, &files_to_parse.files(db));
}
