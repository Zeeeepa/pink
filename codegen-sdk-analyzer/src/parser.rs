use codegen_sdk_common::FileNodeId;
use codegen_sdk_cst::CSTLanguage;
use codegen_sdk_macros::{languages_ast, parse_language};
languages_ast!();

#[salsa::tracked(return_ref)]
pub fn parse_file<'db>(
    db: &'db dyn codegen_sdk_resolution::Db,
    file: codegen_sdk_common::FileNodeId,
) -> Option<ParsedFile<'db>> {
    parse_language!();
    None
}
