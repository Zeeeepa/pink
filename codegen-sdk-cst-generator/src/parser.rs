use std::error::Error;

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Node {
    #[serde(rename = "type")]
    pub type_name: String,
    pub named: bool,
    #[serde(default)]
    pub subtypes: Vec<TypeDefinition>,
    #[serde(default)]
    pub fields: Option<Fields>,
    #[serde(default)]
    pub children: Option<Children>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Fields {
    #[serde(flatten)]
    pub fields: std::collections::HashMap<String, FieldDefinition>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub multiple: bool,
    pub required: bool,
    #[serde(default)]
    pub types: Vec<TypeDefinition>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TypeDefinition {
    #[serde(rename = "type")]
    pub type_name: String,
    pub named: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Children {
    pub multiple: bool,
    pub required: bool,
    #[serde(default)]
    pub types: Vec<TypeDefinition>,
}

pub fn parse_node_types(source: &str) -> Result<Vec<Node>, Box<dyn Error>> {
    let parsed: Vec<Node> = serde_json::from_str(&source)?;
    Ok(parsed)
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_node_types() {
        let source = tree_sitter_python::NODE_TYPES;
        let cst = parse_node_types(source).unwrap();
        panic!();
    }
}
