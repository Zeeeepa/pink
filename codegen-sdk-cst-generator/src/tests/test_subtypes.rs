use std::collections::HashMap;

use assert_tokenstreams_eq::assert_tokenstreams_eq;
use codegen_sdk_common::parser::{Fields, Node, TypeDefinition};
use quote::quote;

use crate::{generate_cst, test_util::get_language};

#[test_log::test]
fn test_basic_subtypes() {
    // Define nodes with basic subtype relationships
    let nodes = vec![
        Node {
            type_name: "expression".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "binary_expression".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "unary_expression".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "binary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "unary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];

    let language = get_language(nodes);
    let output = generate_cst(&language).unwrap();
    let expected = quote! {
        use std::{backtrace::Backtrace, sync::Arc};
        use bytes::Bytes;
        use codegen_sdk_common::*;
        use derive_more::Debug;
        use rkyv::{Archive, Deserialize, Serialize};
        use subenum::subenum;
        use tree_sitter;

        #[derive(Debug, Clone)]
        #[subenum(Expression(derive(Archive, Deserialize, Serialize)))]
        pub enum NodeTypes {
            #[subenum(Expression)]
            BinaryExpression(BinaryExpression),
            #[subenum(Expression)]
            UnaryExpression(UnaryExpression),
        }

        impl std::convert::From<BinaryExpression> for NodeTypes {
            fn from(variant: BinaryExpression) -> Self {
                Self::BinaryExpression(variant)
            }
        }

        impl std::convert::From<UnaryExpression> for NodeTypes {
            fn from(variant: UnaryExpression) -> Self {
                Self::UnaryExpression(variant)
            }
        }

        impl std::convert::From<BinaryExpression> for Expression {
            fn from(variant: BinaryExpression) -> Self {
                Self::BinaryExpression(variant)
            }
        }

        impl std::convert::From<UnaryExpression> for Expression {
            fn from(variant: UnaryExpression) -> Self {
                Self::UnaryExpression(variant)
            }
        }

        impl FromNode for Expression {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "binary_expression" => Ok(Self::BinaryExpression(BinaryExpression::from_node(node, buffer)?)),
                    "unary_expression" => Ok(Self::UnaryExpression(UnaryExpression::from_node(node, buffer)?)),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }
    };
    assert_tokenstreams_eq!(&output, &expected);
}

#[test_log::test]
fn test_nested_subtypes() {
    // Define nodes with nested subtype relationships
    let nodes = vec![
        // Top level statement type
        Node {
            type_name: "statement".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "expression_statement".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Declaration is both a statement subtype and has its own subtypes
        Node {
            type_name: "declaration".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "function_declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "class_declaration".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Concrete node types
        Node {
            type_name: "function_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "class_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "expression_statement".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];

    let language = get_language(nodes);
    let output = generate_cst(&language).unwrap();
    let expected = quote! {
        use std::{backtrace::Backtrace, sync::Arc};
        use bytes::Bytes;
        use codegen_sdk_common::*;
        use derive_more::Debug;
        use rkyv::{Archive, Deserialize, Serialize};
        use subenum::subenum;
        use tree_sitter;

        #[subenum(
            Declaration(derive(Archive, Deserialize, Serialize)),
            Statement(derive(Archive, Deserialize, Serialize))
        )]
        #[derive(Debug, Clone)]
        pub enum NodeTypes {
            #[subenum(Declaration, Statement)]
            ClassDeclaration(ClassDeclaration),
            #[subenum(Statement)]
            ExpressionStatement(ExpressionStatement),
            #[subenum(Declaration, Statement)]
            FunctionDeclaration(FunctionDeclaration),
        }

        impl std::convert::From<ClassDeclaration> for NodeTypes {
            fn from(variant: ClassDeclaration) -> Self {
                Self::ClassDeclaration(variant)
            }
        }

        impl std::convert::From<ClassDeclaration> for Declaration {
            fn from(variant: ClassDeclaration) -> Self {
                Self::ClassDeclaration(variant)
            }
        }

        impl std::convert::From<ExpressionStatement> for NodeTypes {
            fn from(variant: ExpressionStatement) -> Self {
                Self::ExpressionStatement(variant)
            }
        }

        impl std::convert::From<ExpressionStatement> for Statement {
            fn from(variant: ExpressionStatement) -> Self {
                Self::ExpressionStatement(variant)
            }
        }

        impl std::convert::From<FunctionDeclaration> for NodeTypes {
            fn from(variant: FunctionDeclaration) -> Self {
                Self::FunctionDeclaration(variant)
            }
        }

        impl std::convert::From<FunctionDeclaration> for Declaration {
            fn from(variant: FunctionDeclaration) -> Self {
                Self::FunctionDeclaration(variant)
            }
        }

        impl FromNode for Declaration {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "class_declaration" => Ok(Self::ClassDeclaration(ClassDeclaration::from_node(
                        node, buffer,
                    )?)),
                    "function_declaration" => Ok(Self::FunctionDeclaration(
                        FunctionDeclaration::from_node(node, buffer)?,
                    )),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }

        impl FromNode for Statement {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "class_declaration" => Ok(Self::ClassDeclaration(
                        ClassDeclaration::from_node(node, buffer)?,
                    )),
                    "function_declaration" => Ok(Self::FunctionDeclaration(
                        FunctionDeclaration::from_node(node, buffer)?,
                    )),
                    "expression_statement" => Ok(Self::ExpressionStatement(
                        ExpressionStatement::from_node(node, buffer)?,
                    )),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }


        // ... struct definitions and implementations for ClassDeclaration, ExpressionStatement, and FunctionDeclaration ...
    };
    assert_tokenstreams_eq!(&output, &expected);
}

#[test_log::test]
fn test_subtypes_with_fields() {
    let nodes = vec![
        Node {
            type_name: "expression".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "binary_expression".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "literal".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "binary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: Some(Fields {
                fields: HashMap::from([
                    (
                        "left".to_string(),
                        codegen_sdk_common::parser::FieldDefinition {
                            types: vec![TypeDefinition {
                                type_name: "expression".to_string(),
                                named: true,
                            }],
                            multiple: false,
                            required: true,
                        },
                    ),
                    (
                        "right".to_string(),
                        codegen_sdk_common::parser::FieldDefinition {
                            types: vec![TypeDefinition {
                                type_name: "expression".to_string(),
                                named: true,
                            }],
                            multiple: false,
                            required: true,
                        },
                    ),
                ]),
            }),
            children: None,
        },
        Node {
            type_name: "literal".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];

    let language = get_language(nodes);
    let output = generate_cst(&language).unwrap();
    let expected = quote! {
        use std::{backtrace::Backtrace, sync::Arc};
        use bytes::Bytes;
        use codegen_sdk_common::*;
        use derive_more::Debug;
        use rkyv::{Archive, Deserialize, Serialize};
        use subenum::subenum;
        use tree_sitter;
        #[derive(Debug, Clone)]
        pub struct BinaryExpression {
            left: Box<Expression>,
            right: Box<Expression>,
            // ... other required fields ...
        }
        // ... expected impl blocks ...
    };
    assert_tokenstreams_eq!(&output, &expected);
}

#[test_log::test]
fn test_deeply_nested_subtypes() {
    let nodes = vec![
        // Top level statement type
        Node {
            type_name: "statement".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "expression_statement".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Declaration with its subtypes
        Node {
            type_name: "declaration".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "function_declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "class_declaration".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Function declaration with its subtype
        Node {
            type_name: "function_declaration".to_string(),
            subtypes: vec![TypeDefinition {
                type_name: "method_declaration".to_string(),
                named: true,
            }],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Concrete node types
        Node {
            type_name: "method_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "class_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "expression_statement".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];
    let language = get_language(nodes);
    let output = generate_cst(&language).unwrap();
    let expected = quote! {
        use std::{backtrace::Backtrace, sync::Arc};
        use bytes::Bytes;
        use codegen_sdk_common::*;
        use derive_more::Debug;
        use rkyv::{Archive, Deserialize, Serialize};
        use subenum::subenum;
        use tree_sitter;

        #[subenum(
            Declaration(derive(Archive, Deserialize, Serialize)),
            Statement(derive(Archive, Deserialize, Serialize)),
            FunctionDeclaration(derive(Archive, Deserialize, Serialize))
        )]
        #[derive(Debug, Clone)]
        pub enum NodeTypes {
            #[subenum(Declaration, Statement)]
            ClassDeclaration(ClassDeclaration),
            #[subenum(Statement)]
            ExpressionStatement(ExpressionStatement),
            #[subenum(Declaration, Statement, FunctionDeclaration)]
            MethodDeclaration(MethodDeclaration),
        }

        impl std::convert::From<ClassDeclaration> for NodeTypes {
            fn from(variant: ClassDeclaration) -> Self {
                Self::ClassDeclaration(variant)
            }
        }

        impl std::convert::From<ClassDeclaration> for Declaration {
            fn from(variant: ClassDeclaration) -> Self {
                Self::ClassDeclaration(variant)
            }
        }

        impl std::convert::From<ExpressionStatement> for NodeTypes {
            fn from(variant: ExpressionStatement) -> Self {
                Self::ExpressionStatement(variant)
            }
        }

        impl std::convert::From<ExpressionStatement> for Statement {
            fn from(variant: ExpressionStatement) -> Self {
                Self::ExpressionStatement(variant)
            }
        }

        impl std::convert::From<MethodDeclaration> for NodeTypes {
            fn from(variant: MethodDeclaration) -> Self {
                Self::MethodDeclaration(variant)
            }
        }

        impl std::convert::From<MethodDeclaration> for Declaration {
            fn from(variant: MethodDeclaration) -> Self {
                Self::MethodDeclaration(variant)
            }
        }

        impl std::convert::From<MethodDeclaration> for FunctionDeclaration {
            fn from(variant: MethodDeclaration) -> Self {
                Self::MethodDeclaration(variant)
            }
        }

        impl FromNode for Declaration {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "class_declaration" => Ok(Self::ClassDeclaration(ClassDeclaration::from_node(
                        node, buffer,
                    )?)),
                    "method_declaration" => Ok(Self::MethodDeclaration(
                        MethodDeclaration::from_node(node, buffer)?,
                    )),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }

        impl FromNode for Statement {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "class_declaration" => Ok(Self::ClassDeclaration(
                        ClassDeclaration::from_node(node, buffer)?,
                    )),
                    "method_declaration" => Ok(Self::MethodDeclaration(
                        MethodDeclaration::from_node(node, buffer)?,
                    )),
                    "expression_statement" => Ok(Self::ExpressionStatement(
                        ExpressionStatement::from_node(node, buffer)?,
                    )),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }

        impl FromNode for FunctionDeclaration {
            fn from_node(node: tree_sitter::Node, buffer: &Arc<Bytes>) -> Result<Self, ParseError> {
                match node.kind() {
                    "method_declaration" => Ok(Self::MethodDeclaration(
                        MethodDeclaration::from_node(node, buffer)?,
                    )),
                    _ => Err(ParseError::UnexpectedNode {
                        node_type: node.kind().to_string(),
                        backtrace: Backtrace::capture(),
                    }),
                }
            }
        }

        // ... struct definitions and implementations for ClassDeclaration, ExpressionStatement, and MethodDeclaration ...
    };
    assert_tokenstreams_eq!(&output, &expected);
}
