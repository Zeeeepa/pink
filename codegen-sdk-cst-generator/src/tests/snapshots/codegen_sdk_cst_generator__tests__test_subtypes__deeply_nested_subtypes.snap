---
source: codegen-sdk-cst-generator/src/tests/test_subtypes.rs
expression: "crate::test_util::snapshot_string(&output)"
---
use std::sync::Arc;
use tree_sitter;
use codegen_sdk_common::*;
use subenum::subenum;
use std::backtrace::Backtrace;
use bytes::Bytes;
use derive_generic_visitor::Drive;
use ambassador::Delegate;
use codegen_sdk_cst::CSTLanguage;
use codegen_sdk_common::language::Language;
#[subenum(Declaration, FunctionDeclaration, Statement)]
#[derive(Debug, Clone, Eq, PartialEq, Drive, Hash, salsa::Update, Delegate)]
#[delegate(CSTNode<'db1>)]
pub enum NodeTypes<'db1> {
    #[subenum(Declaration, Statement)]
    ClassDeclaration(ClassDeclaration<'db1>),
    #[subenum(Statement)]
    ExpressionStatement(ExpressionStatement<'db1>),
    #[subenum(FunctionDeclaration, Declaration, Statement)]
    MethodDeclaration(MethodDeclaration<'db1>),
}
impl<'db3> From<ClassDeclaration<'db3>> for NodeTypes<'db3> {
    fn from(node: ClassDeclaration<'db3>) -> Self {
        Self::ClassDeclaration(node)
    }
}
impl<'db3> From<ExpressionStatement<'db3>> for NodeTypes<'db3> {
    fn from(node: ExpressionStatement<'db3>) -> Self {
        Self::ExpressionStatement(node)
    }
}
impl<'db3> From<MethodDeclaration<'db3>> for NodeTypes<'db3> {
    fn from(node: MethodDeclaration<'db3>) -> Self {
        Self::MethodDeclaration(node)
    }
}
impl<'db4> FromNode<'db4> for Declaration<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => {
                Ok(
                    Self::MethodDeclaration(
                        MethodDeclaration::from_node(db, node, buffer)?,
                    ),
                )
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4> for FunctionDeclaration<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => {
                Ok(
                    Self::MethodDeclaration(
                        MethodDeclaration::from_node(db, node, buffer)?,
                    ),
                )
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4> for Statement<'db4> {
    fn from_node(
        db: &'db4 dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        match node.kind_id() {
            0u16 => {
                Ok(
                    Self::MethodDeclaration(
                        MethodDeclaration::from_node(db, node, buffer)?,
                    ),
                )
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct ClassDeclaration<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
}
impl<'db> FromNode<'db> for ClassDeclaration<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
        })
    }
}
impl<'db> CSTNode<'db> for ClassDeclaration<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for ClassDeclaration<'db1> {
    type Child = Self;
    fn children(&self) -> Vec<Self::Child> {
        vec![]
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for ClassDeclaration<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct ExpressionStatement<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
}
impl<'db> FromNode<'db> for ExpressionStatement<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
        })
    }
}
impl<'db> CSTNode<'db> for ExpressionStatement<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for ExpressionStatement<'db1> {
    type Child = Self;
    fn children(&self) -> Vec<Self::Child> {
        vec![]
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for ExpressionStatement<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[derive(Debug, Clone, Drive, Eq, PartialEq, salsa::Update)]
pub struct MethodDeclaration<'db> {
    #[drive(skip)]
    start_byte: usize,
    #[drive(skip)]
    end_byte: usize,
    #[drive(skip)]
    _kind: std::string::String,
    #[drive(skip)]
    start_position: Point<'db>,
    #[drive(skip)]
    end_position: Point<'db>,
    #[drive(skip)]
    buffer: Arc<Bytes>,
    #[drive(skip)]
    kind_id: u16,
    #[drive(skip)]
    is_error: bool,
    #[drive(skip)]
    named: bool,
    #[drive(skip)]
    id: usize,
}
impl<'db> FromNode<'db> for MethodDeclaration<'db> {
    fn from_node(
        db: &'db dyn salsa::Database,
        node: tree_sitter::Node,
        buffer: &Arc<Bytes>,
    ) -> Result<Self, ParseError> {
        let start_position = Point::from(db, node.start_position());
        let end_position = Point::from(db, node.end_position());
        Ok(Self {
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
            _kind: node.kind().to_string(),
            start_position: start_position,
            end_position: end_position,
            buffer: buffer.clone(),
            kind_id: node.kind_id(),
            is_error: node.is_error(),
            named: node.is_named(),
            id: node.id(),
        })
    }
}
impl<'db> CSTNode<'db> for MethodDeclaration<'db> {
    fn kind(&self) -> &str {
        &self._kind
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        self.kind_id
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        self.named
    }
    fn id(&self) -> usize {
        self.id
    }
}
impl<'db1> HasChildren<'db1> for MethodDeclaration<'db1> {
    type Child = Self;
    fn children(&self) -> Vec<Self::Child> {
        vec![]
    }
    fn children_by_field_name(&self, field_name: &str) -> Vec<Self::Child> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id(&self, field_id: u16) -> Vec<Self::Child> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for MethodDeclaration<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
#[salsa::tracked]
pub struct Parsed<'db> {
    #[tracked]
    #[return_ref]
    pub program: Option<Program<'db>>,
}
pub fn parse_program_raw(
    db: &dyn salsa::Database,
    input: codegen_sdk_cst::Input,
) -> Option<Program<'_>> {
    let buffer = Bytes::from(input.content(db).as_bytes().to_vec());
    let tree = codegen_sdk_common::language::language::Language
        .parse_tree_sitter(&input.content(db));
    match tree {
        Ok(tree) => {
            if tree.root_node().has_error() {
                ParseError::SyntaxError.report(db);
                None
            } else {
                let buffer = Arc::new(buffer);
                Program::from_node(db, tree.root_node(), &buffer)
                    .map_or_else(
                        |e| {
                            e.report(db);
                            None
                        },
                        |program| { Some(program) },
                    )
            }
        }
        Err(e) => {
            e.report(db);
            None
        }
    }
}
#[salsa::tracked]
pub fn parse_program(
    db: &dyn salsa::Database,
    input: codegen_sdk_cst::Input,
) -> Parsed<'_> {
    Parsed::new(db, parse_program_raw(db, input))
}
pub struct Language;
impl CSTLanguage for Language {
    type Program<'db> = Program<'db>;
    fn language() -> &'static Language {
        &codegen_sdk_common::language::language::Language
    }
    fn parse<'db>(
        db: &'db dyn salsa::Database,
        content: std::string::String,
    ) -> &'db Option<Self::Program<'db>> {
        let input = codegen_sdk_cst::Input::new(db, content);
        return parse_program(db, input).program(db);
    }
}
